CODING_STYLE.md

```diff
@@ -112,10 +112,9 @@ Use `solAssert` and `solUnimplementedAssert` generously to check assumptions tha
 6. If a function returns multiple values, use std::tuple (std::pair acceptable) or better introduce a struct type. Do not use */& arguments.
 7. Use parameters of pointer type only if ``nullptr`` is a valid argument, use references otherwise. Often, ``std::optional`` is better suited than a raw pointer.
 8. Never use a macro where adequate non-preprocessor C++ can be written.
-9. Only use ``auto`` if the type is very long and rather irrelevant.
```

Comment:
I'd not be opposed to relaxing this rule, but I think that just outright removing it is a bad idea. We should still limit the use of `auto` somewhat.

The thing that I dislike about `auto` the most is that it optimizes for writing code rather than reading it and for writing it's not even a meaningful speedup because modern completion and IDE features make that already easy. On the other hand it shifts the effort of figuring out the type from the one who writes the code to the one who reads it. Even if the type can be seen from the nearby context, having to fish it out slows down reading. It's only really redundant if the type is obvious directly from the expression it's used on.

Reply:
I wholeheartedly disagree here. Whenever you *can't* read the code due to an `auto`, there's something wrong with it other than the `auto` - it doesn't allow you to skip types, if they *are* relevant, since you still have type-checking - in all cases in which you *can* use ``auto``, it's better to use ``auto``, since the code should be apparent as correct despite not knowing the respective type. What slows down reading code is having to wonder about pointer-, reference- or const-ness of things, so those I'd usually preserve despite using `auto`.
I can bring the point back in the way that's my preferred style for contrast :-).

Reply:
But that's just to make the point :-). You're reviewing more code than I these days, so I'd generally assign more value to your preference here in the end.

Reply:
I'd actually even argue the contrary: *not* using `auto` makes code harder to read, since then I actually need to check, if the type is correct and whether there was an implicit conversion or the like, while *if* I use ``auto``, I do know all relevant aspects of the type, despite it not being written explicitly.

Reply:
Well, having an example to argue about would probably be helpful. Maybe I should try to come up with one too.

Using wrong type can be bad if it compiles but more often than not it will not. It will instead let compiler catch your bad assumption.

Overall I do think we should relax this but I think what you propose is a bit too extreme :) But I'll come back to this with more concrete arguments after the meeting.

Reply:
Yeah, no worries - we can also collect more opinions in the next weeks :-). And don't need to be too extreme with it. But yeah: to argue the point: the compiler will, in a lot of cases, *accept* wrong types, and perform implicit conversions, which may not what you want at all.

Also: official recommendations are on my side: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es11-use-auto-to-avoid-redundant-repetition-of-type-names